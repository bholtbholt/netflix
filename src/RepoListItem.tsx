import * as React from 'react';
import { Repo } from './github';
import { GitPullRequestIcon, IssueOpenedIcon, StarIcon } from '@primer/octicons-react';

export const RepoListItem: React.FC<Repo> = ({
  description,
  html_url,
  language,
  name,
  open_issues_count,
  stargazers_count,
}) => {
  return (
    <li>
      <h5>
        <a href={html_url} rel="nofollow" target="_blank">
          {name}
        </a>
      </h5>
      {description && <p>{description}</p>}
      <ul>
        <li>
          <a href={`${html_url}/pulls`} rel="nofollow" target="_blank" aria-label="Pull requests">
            <GitPullRequestIcon />
            PRs
          </a>
        </li>
        <li>
          <a href={`${html_url}/issues`} rel="nofollow" target="_blank">
            <IssueOpenedIcon aria-label="Open issues" />
            {open_issues_count}
          </a>
        </li>
        <li>
          <a href={`${html_url}/stargazers`} rel="nofollow" target="_blank">
            <StarIcon aria-label="Star count" />
            {stargazers_count}
          </a>
        </li>
        <li>{language}</li>
      </ul>
    </li>
  );
};
