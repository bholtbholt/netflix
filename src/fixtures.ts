import { AuthorOrCommitter, Commit, CommitDetails, Repo } from './github';

export function createRepo(repo = {}): Repo {
  const mockRepo: Repo = {
    id: 8759133,
    node_id: 'MDEwOlJlcG9zaXRvcnk4NzU5MTMz',
    name: 'zuul',
    full_name: 'Netflix/zuul',
    private: false,
    owner: {
      login: 'Netflix',
      id: 913567,
      node_id: 'MDEyOk9yZ2FuaXphdGlvbjkxMzU2Nw==',
      avatar_url: 'https://avatars.githubusercontent.com/u/913567?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/Netflix',
      html_url: 'https://github.com/Netflix',
      followers_url: 'https://api.github.com/users/Netflix/followers',
      following_url: 'https://api.github.com/users/Netflix/following{/other_user}',
      gists_url: 'https://api.github.com/users/Netflix/gists{/gist_id}',
      starred_url: 'https://api.github.com/users/Netflix/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/Netflix/subscriptions',
      organizations_url: 'https://api.github.com/users/Netflix/orgs',
      repos_url: 'https://api.github.com/users/Netflix/repos',
      events_url: 'https://api.github.com/users/Netflix/events{/privacy}',
      received_events_url: 'https://api.github.com/users/Netflix/received_events',
      type: 'Organization',
      site_admin: false,
    },
    html_url: 'https://github.com/Netflix/zuul',
    description:
      'Zuul is a gateway service that provides dynamic routing, monitoring, resiliency, security, and more.',
    fork: false,
    url: 'https://api.github.com/repos/Netflix/zuul',
    forks_url: 'https://api.github.com/repos/Netflix/zuul/forks',
    keys_url: 'https://api.github.com/repos/Netflix/zuul/keys{/key_id}',
    collaborators_url: 'https://api.github.com/repos/Netflix/zuul/collaborators{/collaborator}',
    teams_url: 'https://api.github.com/repos/Netflix/zuul/teams',
    hooks_url: 'https://api.github.com/repos/Netflix/zuul/hooks',
    issue_events_url: 'https://api.github.com/repos/Netflix/zuul/issues/events{/number}',
    events_url: 'https://api.github.com/repos/Netflix/zuul/events',
    assignees_url: 'https://api.github.com/repos/Netflix/zuul/assignees{/user}',
    branches_url: 'https://api.github.com/repos/Netflix/zuul/branches{/branch}',
    tags_url: 'https://api.github.com/repos/Netflix/zuul/tags',
    blobs_url: 'https://api.github.com/repos/Netflix/zuul/git/blobs{/sha}',
    git_tags_url: 'https://api.github.com/repos/Netflix/zuul/git/tags{/sha}',
    git_refs_url: 'https://api.github.com/repos/Netflix/zuul/git/refs{/sha}',
    trees_url: 'https://api.github.com/repos/Netflix/zuul/git/trees{/sha}',
    statuses_url: 'https://api.github.com/repos/Netflix/zuul/statuses/{sha}',
    languages_url: 'https://api.github.com/repos/Netflix/zuul/languages',
    stargazers_url: 'https://api.github.com/repos/Netflix/zuul/stargazers',
    contributors_url: 'https://api.github.com/repos/Netflix/zuul/contributors',
    subscribers_url: 'https://api.github.com/repos/Netflix/zuul/subscribers',
    subscription_url: 'https://api.github.com/repos/Netflix/zuul/subscription',
    commits_url: 'https://api.github.com/repos/Netflix/zuul/commits{/sha}',
    git_commits_url: 'https://api.github.com/repos/Netflix/zuul/git/commits{/sha}',
    comments_url: 'https://api.github.com/repos/Netflix/zuul/comments{/number}',
    issue_comment_url: 'https://api.github.com/repos/Netflix/zuul/issues/comments{/number}',
    contents_url: 'https://api.github.com/repos/Netflix/zuul/contents/{+path}',
    compare_url: 'https://api.github.com/repos/Netflix/zuul/compare/{base}...{head}',
    merges_url: 'https://api.github.com/repos/Netflix/zuul/merges',
    archive_url: 'https://api.github.com/repos/Netflix/zuul/{archive_format}{/ref}',
    downloads_url: 'https://api.github.com/repos/Netflix/zuul/downloads',
    issues_url: 'https://api.github.com/repos/Netflix/zuul/issues{/number}',
    pulls_url: 'https://api.github.com/repos/Netflix/zuul/pulls{/number}',
    milestones_url: 'https://api.github.com/repos/Netflix/zuul/milestones{/number}',
    notifications_url:
      'https://api.github.com/repos/Netflix/zuul/notifications{?since,all,participating}',
    labels_url: 'https://api.github.com/repos/Netflix/zuul/labels{/name}',
    releases_url: 'https://api.github.com/repos/Netflix/zuul/releases{/id}',
    deployments_url: 'https://api.github.com/repos/Netflix/zuul/deployments',
    created_at: '2013-03-13T18:53:57Z',
    updated_at: '2021-09-11T16:32:45Z',
    pushed_at: '2021-09-01T18:55:22Z',
    git_url: 'git://github.com/Netflix/zuul.git',
    ssh_url: 'git@github.com:Netflix/zuul.git',
    clone_url: 'https://github.com/Netflix/zuul.git',
    svn_url: 'https://github.com/Netflix/zuul',
    homepage: '',
    size: 4224,
    stargazers_count: 11271,
    watchers_count: 11271,
    language: 'Java',
    has_issues: true,
    has_projects: true,
    has_downloads: true,
    has_wiki: true,
    has_pages: true,
    forks_count: 2110,
    mirror_url: null,
    archived: false,
    disabled: false,
    open_issues_count: 251,
    license: {
      key: 'apache-2.0',
      name: 'Apache License 2.0',
      spdx_id: 'Apache-2.0',
      url: 'https://api.github.com/licenses/apache-2.0',
      node_id: 'MDc6TGljZW5zZTI=',
    },
    forks: 2110,
    open_issues: 251,
    watchers: 11271,
    default_branch: 'master',
    permissions: {
      admin: false,
      push: false,
      pull: true,
    },
  };

  return { ...mockRepo, ...repo };
}

export function createCommit(commit = {}): Commit {
  const mockCommitDetails: CommitDetails = {
    author: {
      name: 'Carl Mastrangelo',
      email: 'carl-mastrangelo@users.noreply.github.com',
      date: '2021-09-01T18:55:22Z',
    },
    committer: {
      name: 'GitHub',
      email: 'noreply@github.com',
      date: '2021-09-01T18:55:22Z',
    },
    message: 'core: include proxy protocol in server conn dimms',
    tree: {
      sha: '138126c9ea71224b827889df4abbab5fd148bb48',
      url: 'https://api.github.com/repos/Netflix/zuul/git/trees/138126c9ea71224b827889df4abbab5fd148bb48',
    },
    url: 'https://api.github.com/repos/Netflix/zuul/git/commits/6d29e74fd27e6e3ad8b7efa8de511a567654cf63',
    comment_count: 0,
    verification: {
      verified: true,
      reason: 'valid',
    },
  };

  const mockAuthor: AuthorOrCommitter = {
    login: 'carl-mastrangelo',
    id: 8943572,
    node_id: 'MDQ6VXNlcjg5NDM1NzI=',
    avatar_url: 'https://avatars.githubusercontent.com/u/8943572?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/carl-mastrangelo',
    html_url: 'https://github.com/carl-mastrangelo',
    followers_url: 'https://api.github.com/users/carl-mastrangelo/followers',
    following_url: 'https://api.github.com/users/carl-mastrangelo/following{/other_user}',
    gists_url: 'https://api.github.com/users/carl-mastrangelo/gists{/gist_id}',
    starred_url: 'https://api.github.com/users/carl-mastrangelo/starred{/owner}{/repo}',
    subscriptions_url: 'https://api.github.com/users/carl-mastrangelo/subscriptions',
    organizations_url: 'https://api.github.com/users/carl-mastrangelo/orgs',
    repos_url: 'https://api.github.com/users/carl-mastrangelo/repos',
    events_url: 'https://api.github.com/users/carl-mastrangelo/events{/privacy}',
    received_events_url: 'https://api.github.com/users/carl-mastrangelo/received_events',
    type: 'User',
    site_admin: false,
  };
  const mockCommit: Commit = {
    sha: '6d29e74fd27e6e3ad8b7efa8de511a567654cf63',
    node_id: 'MDY6Q29tbWl0ODc1OTEzMzo2ZDI5ZTc0ZmQyN2U2ZTNhZDhiN2VmYThkZTUxMWE1Njc2NTRjZjYz',
    commit: mockCommitDetails,
    url: 'https://api.github.com/repos/Netflix/zuul/commits/6d29e74fd27e6e3ad8b7efa8de511a567654cf63',
    html_url: 'https://github.com/Netflix/zuul/commit/6d29e74fd27e6e3ad8b7efa8de511a567654cf63',
    comments_url:
      'https://api.github.com/repos/Netflix/zuul/commits/6d29e74fd27e6e3ad8b7efa8de511a567654cf63/comments',
    author: mockAuthor,
    committer: mockAuthor,
  };

  return { ...mockCommit, ...commit };
}
