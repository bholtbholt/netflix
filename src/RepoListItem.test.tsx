import * as React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { RepoListItem } from './RepoListItem';
import { createCommit, createRepo } from './fixtures';

beforeEach(() => {
  jest.resetAllMocks();
  jest.spyOn(window, 'fetch');
  (window.fetch as jest.Mock).mockReturnValue({
    ok: true,
    json: async () => [createCommit()],
  });
});

it('should match snapshot', async () => {
  const snapshot = renderer.create(<RepoListItem {...createRepo()} />).toJSON();

  expect(snapshot).toMatchSnapshot();
});

it('should request recent commits', async () => {
  render(<RepoListItem {...createRepo()} />);

  const button = screen.getByText('Load latest commits') as HTMLButtonElement;

  await fireEvent.click(button);

  expect(window.fetch).toHaveBeenLastCalledWith(
    'https://api.github.com/repos/Netflix/zuul/commits?per_page=10',
    expect.objectContaining({}),
  );
});

it('should render recent commits to the repo container', async () => {
  const commitFixture = createCommit();
  const commitMessage = commitFixture.commit.message;

  render(<RepoListItem {...createRepo()} />);

  let commit = screen.queryByText(commitMessage);
  expect(commit).not.toBeInTheDocument();

  const button = screen.getByText('Load latest commits') as HTMLButtonElement;

  await fireEvent.click(button);
  commit = await screen.findByText(commitMessage);

  expect(commit).toBeInTheDocument();
});
