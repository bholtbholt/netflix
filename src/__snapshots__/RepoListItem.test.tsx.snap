// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<li
  className="p-6 mx-auto bg-white rounded-xl shadow-xl mb-3"
  id="zuul"
>
  <h5
    className="text-3xl font-bold text-gray-900"
  >
    <a
      className="hover:text-indigo-600"
      href="https://github.com/Netflix/zuul"
      rel="nofollow"
      target="_blank"
    >
      zuul
    </a>
  </h5>
  <p
    className="text-gray-500 text-xl"
  >
    Zuul is a gateway service that provides dynamic routing, monitoring, resiliency, security, and more.
  </p>
  <ul
    className="flex items-center text-gray-600 mt-6 mb-8 gap-x-8"
  >
    <li
      className="flex-initial"
    >
      <a
        aria-label="Pull requests"
        className="rounded-sm hover:text-indigo-800 hover:underline focus:text-indigo-800 focus:outline-none focus:ring-2 focus:ring-offset-8 focus:ring-indigo-500 transition duration-200 ease-in-out"
        href="https://github.com/Netflix/zuul/pulls"
        rel="nofollow"
        target="_blank"
      >
        <svg
          aria-hidden="true"
          className="mr-2"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path fill-rule=\\"evenodd\\" d=\\"M7.177 3.073L9.573.677A.25.25 0 0110 .854v4.792a.25.25 0 01-.427.177L7.177 3.427a.25.25 0 010-.354zM3.75 2.5a.75.75 0 100 1.5.75.75 0 000-1.5zm-2.25.75a2.25 2.25 0 113 2.122v5.256a2.251 2.251 0 11-1.5 0V5.372A2.25 2.25 0 011.5 3.25zM11 2.5h-1V4h1a1 1 0 011 1v5.628a2.251 2.251 0 101.5 0V5A2.5 2.5 0 0011 2.5zm1 10.25a.75.75 0 111.5 0 .75.75 0 01-1.5 0zM3.75 12a.75.75 0 100 1.5.75.75 0 000-1.5z\\"></path>",
            }
          }
          fill="currentColor"
          height={16}
          role="img"
          style={
            Object {
              "display": "inline-block",
              "overflow": "visible",
              "userSelect": "none",
              "verticalAlign": "text-bottom",
            }
          }
          viewBox="0 0 16 16"
          width={16}
        />
        PRs
      </a>
    </li>
    <li
      className="flex-initial"
    >
      <a
        className="rounded-sm hover:text-indigo-800 hover:underline focus:text-indigo-800 focus:outline-none focus:ring-2 focus:ring-offset-8 focus:ring-indigo-500 transition duration-200 ease-in-out"
        href="https://github.com/Netflix/zuul/issues"
        rel="nofollow"
        target="_blank"
      >
        <svg
          aria-hidden="false"
          aria-label="Open issues"
          className="mr-2"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path d=\\"M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z\\"></path><path fill-rule=\\"evenodd\\" d=\\"M8 0a8 8 0 100 16A8 8 0 008 0zM1.5 8a6.5 6.5 0 1113 0 6.5 6.5 0 01-13 0z\\"></path>",
            }
          }
          fill="currentColor"
          height={16}
          role="img"
          style={
            Object {
              "display": "inline-block",
              "overflow": "visible",
              "userSelect": "none",
              "verticalAlign": "text-bottom",
            }
          }
          viewBox="0 0 16 16"
          width={16}
        />
        251
      </a>
    </li>
    <li
      className="flex-initial"
    >
      <a
        className="rounded-sm hover:text-indigo-800 hover:underline focus:text-indigo-800 focus:outline-none focus:ring-2 focus:ring-offset-8 focus:ring-indigo-500 transition duration-200 ease-in-out"
        href="https://github.com/Netflix/zuul/stargazers"
        rel="nofollow"
        target="_blank"
      >
        <svg
          aria-hidden="false"
          aria-label="Star count"
          className="mr-2"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path fill-rule=\\"evenodd\\" d=\\"M8 .25a.75.75 0 01.673.418l1.882 3.815 4.21.612a.75.75 0 01.416 1.279l-3.046 2.97.719 4.192a.75.75 0 01-1.088.791L8 12.347l-3.766 1.98a.75.75 0 01-1.088-.79l.72-4.194L.818 6.374a.75.75 0 01.416-1.28l4.21-.611L7.327.668A.75.75 0 018 .25zm0 2.445L6.615 5.5a.75.75 0 01-.564.41l-3.097.45 2.24 2.184a.75.75 0 01.216.664l-.528 3.084 2.769-1.456a.75.75 0 01.698 0l2.77 1.456-.53-3.084a.75.75 0 01.216-.664l2.24-2.183-3.096-.45a.75.75 0 01-.564-.41L8 2.694v.001z\\"></path>",
            }
          }
          fill="currentColor"
          height={16}
          role="img"
          style={
            Object {
              "display": "inline-block",
              "overflow": "visible",
              "userSelect": "none",
              "verticalAlign": "text-bottom",
            }
          }
          viewBox="0 0 16 16"
          width={16}
        />
        11271
      </a>
    </li>
    <li
      className="flex-initial"
    >
      <span
        className="px-3 inline-flex text-sm font-semibold rounded-full bg-indigo-100 text-indigo-800"
      >
        Java
      </span>
    </li>
  </ul>
  <div
    className="pl-4"
  >
    <ul
      className="border-l-4 border-gray-100 pl-3"
      id="zuul-commits"
    />
  </div>
  <button
    className="bg-white py-2 px-4 border border-indigo-300 rounded-full shadow-sm text-sm leading-4 font-medium text-indigo-800 hover:bg-indigo-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 transition duration-200 ease-in-out disabled:opacity-50 inline-flex justify-center items-center"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Load latest commits
  </button>
</li>
`;
